src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+            // This means that in general we could have diagnostics produced by one generated file that map to many different actual razor files.
+            // We can't filter them out as we don't know which razor file(s) the underlying generated C# document actually maps to, and which are just imported.
+            // So we publish them all and let them get de-duped.
+            var fileUriToDiagnostics = diagnostics.GroupBy(diagnostic => GetDiagnosticUri(document, diagnostic)).ToDictionary(
+                group => group.Key,
+                group => group.Select(diagnostic => ConvertToLspDiagnostic(diagnostic, text)).ToArray());
+            return fileUriToDiagnostics;
+
+            static Uri GetDiagnosticUri(Document document, DiagnosticData diagnosticData)
             {
-                Code = diagnostic.Id,
-                Message = diagnostic.Message,
-                Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnostic.Severity),
-                Range = ProtocolConversions.TextSpanToRange(DiagnosticData.GetExistingOrCalculatedTextSpan(diagnostic.DataLocation, text), text),
-                // Only the unnecessary diagnostic tag is currently supported via LSP.
-                Tags = diagnostic.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
-            }).ToArray();
+                var filePath = diagnosticData.DataLocation?.MappedFilePath ?? document.FilePath;
+                return ProtocolConversions.GetUriFromFilePath(filePath);
+            }
+
+            static LanguageServer.Protocol.Diagnostic ConvertToLspDiagnostic(DiagnosticData diagnosticData, SourceText text)
+            {
+                return new LanguageServer.Protocol.Diagnostic
+                {
+                    Code = diagnosticData.Id,
+                    Message = diagnosticData.Message,
+                    Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnosticData.Severity),
+                    Range = GetDiagnosticRange(diagnosticData.DataLocation, text),
+                    // Only the unnecessary diagnostic tag is currently supported via LSP.
+                    Tags = diagnosticData.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
+                };
+            }
+        }
+
+        private bool IncludeDiagnostic(DiagnosticData diagnostic)
+        {
+            if (!diagnostic.Properties.TryGetValue(nameof(DocumentPropertiesService.DiagnosticsLspClientName), out var diagnosticClientName))
+            {
+                // This diagnostic is not restricted to a specific LSP client, so just pass it through
+                return true;
+            }
+
+            // We only include this diagnostic if it directly matches our name.
+            return diagnosticClientName == _clientName;
```

Comment:
this seems so fragile.

Reply:
@CyrusNajmabadi I'm all ears for a better suggestion; this is the best we all could come up with...